source/extensions/quic_listeners/quiche/platform/flags_impl.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <string>
+#include <unordered_map>
+
+#include "absl/synchronization/mutex.h"
+
+namespace quiche {
+
+class Flag;
+
+// Registry of QUICHE flags. Can be used to reset all flags to default values,
+// and to look up and set flags by name.
+class FlagRegistry {
+public:
+  ~FlagRegistry() = default;
+
+  // Return singleton instance.
+  static FlagRegistry& GetInstance();
+
+  // Add a flag to the registry.
+  void RegisterFlag(const char* name, Flag* flag);
+
+  // Reset all registered flags to their default values.
+  void ResetFlags() const;
+
+  // Look up a flag by name.
+  Flag* FindFlag(const std::string& name) const;
+
+private:
+  FlagRegistry() = default;
+
+  mutable absl::Mutex mutex_;
+  std::unordered_map<std::string, Flag*> flags_ GUARDED_BY(mutex_);
```

Comment:
The flags are read pretty often, how about use absl::flat_hash_map here for faster lookups?

Reply:
Done. Note that ordinary reads of flags from QUICHE code (e.g. via Get(Quic|Spdy|Http2)(Reloadable|Restart)Flag) access the quiche::FLAGS_* field directly, and do not go through FlagRegistry. FlagRegistry is only involved when looking up flags by std::string name.